File: simp-api-recipes/schemas/cauldron.py
from pydantic import BaseModel
from datetime import datetime
from uuid import UUID
from typing import Optional

# -------------------------
# Cauldron Schemas
# -------------------------

class CauldronBase(BaseModel):
    user_id: UUID
    recipe_id: UUID
    is_active: Optional[bool] = True

class CauldronCreate(CauldronBase):
    """
    Schema for creating a new Cauldron entry.
    """
    pass

class CauldronUpdate(BaseModel):
    is_active: Optional[bool] = None

    class Config:
        from_attributes = True


class Cauldron(CauldronBase):
    """
    Schema for reading a Cauldron entry from the DB.
    """
    cauldron_id: UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

# -------------------------
# CauldronData Schemas
# -------------------------

class CauldronDataBase(BaseModel):
    usage_count: Optional[int] = 0
    last_used: Optional[datetime] = None
    overall_rating: Optional[float] = None
    taste_rating: Optional[float] = None
    ease_rating: Optional[float] = None

class CauldronDataCreate(CauldronDataBase):
    """
    Schema for creating a new CauldronData entry.
    """
    cauldron_id: UUID

class CauldronData(CauldronDataBase):
    """
    Schema for reading CauldronData from the DB.
    """
    cauldron_data_id: UUID
    updated_at: datetime

    class Config:
        from_attributes = True

# -------------------------
# RecipeAnalytics Schemas
# -------------------------

class RecipeAnalyticsBase(BaseModel):
    recipe_id: UUID
    minimum_price: float
    total_calories: Optional[float] = None

class RecipeAnalyticsCreate(RecipeAnalyticsBase):
    """
    Schema for creating a new RecipeAnalytics entry.
    """
    pass

class RecipeAnalytics(RecipeAnalyticsBase):
    """
    Schema for reading RecipeAnalytics from the DB.
    """
    recipe_analytics_id: UUID
    updated_at: datetime
    created_at: datetime

    class Config:
        from_attributes = True

File: simp-front-end/src/lib/types/cauldron.ts
// types/cauldron.ts

// Main Cauldron type: associates a user with a recipe and marks if it's active in the user's rotation.
export type Cauldron = {
    cauldron_id: string;
    user_id: string;
    recipe_id: string;
    is_active: boolean;
    created_at: string;
    updated_at: string;
  };
  
  // For creating a new cauldron record.
  export type CauldronCreate = {
    user_id: string;
    recipe_id: string;
    is_active?: boolean;  // Optional; defaults to true on the backend.
  };
  
  // For updating an existing cauldron record.
  export type CauldronUpdate = Partial<CauldronCreate>;
  
  
  // CauldronData holds dynamic, user-specific usage metrics.
  export type CauldronData = {
    cauldron_data_id: string;
    cauldron_id: string;
    usage_count: number;
    last_used?: string;
    overall_rating?: number;
    taste_rating?: number;
    ease_rating?: number;
    updated_at: string;
  };
  
  // For creating a new cauldron data record.
  export type CauldronDataCreate = {
    cauldron_id: string;
    usage_count?: number;
    last_used?: string;
    overall_rating?: number;
    taste_rating?: number;
    ease_rating?: number;
  };

  export type CauldronRecipe = {
    cauldron_id: string;
    user_id: string;
    recipe_id: string;
    is_active: boolean;
    created_at: string;
    updated_at: string;
    title: string;
    tags: string[];
    front_image: string;
  };
  
  
  // For updating an existing cauldron data record.
  export type CauldronDataUpdate = Partial<CauldronDataCreate>;
  

  